sprs/src/num_kinds.rs

```diff
@@ -2,13 +2,106 @@
 //! or a complex.
 
 use num_complex::{Complex32, Complex64};
-use std::fmt;
+use num_traits::{Num, NumCast, One, ToPrimitive, Zero};
+use std::{
+    fmt,
+    ops::{Add, Div, Mul, Rem, Sub},
+};
+/// the type for Pattern data, it's special which contains no data
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct Pattern;
+
+impl Add for Pattern {
+    type Output = Pattern;
+    fn add(self, _other: Pattern) -> Pattern {
+        Pattern {}
+    }
+}
+impl Zero for Pattern {
+    fn zero() -> Self {
+        Pattern {}
+    }
+    fn is_zero(&self) -> bool {
+        true
+    }
+
+    fn set_zero(&mut self) {
+        *self = Zero::zero();
+    }
+}
+impl Rem for Pattern {
```

Comment:
I am not sure why we need these traits, they make little sense for a pattern type

Reply:
Currently the to_csr and to_csc method needs <N> to be Num. Those traits are required to implement Num.
But in to_csr and to_csc, only Add and Zero are used. I suggest removing the Sum trait requirement for N. We can use N: Add<Output=N> instead. please check the newest Commit for detailed changes.

Reply:
The removal of the Num Trait .and then adding a new Add trait doesn't import break change. Cause all types implement Num must already implement Add.